### PR TITLE
Add ResNet example

### LINKED ISSUE #7: Add the ResNet example
The ResNet is one of [the first wave models](https://github.com/webmachinelearning/webnn/blob/master/op_compatibility/first_wave_models.md).

### PR DESCRIPTION
**Pull request by Honry**
Supports ResNet 101 V2 from TFLite and ResNet 50 V2 from ONNX.

Fixed #7.

### CODE REVIEW

**Comment by Honry** [comment]
@huningxin, PTAL, thanks! I will add the preview once the corresponding test-data [PR](https://github.com/webmachinelearning/test-data/pull/5) landed.

**Comment by huningxin** [comment]
test data PR merged @Honry

**Comment by Honry** [comment]
@huningxin, thanks! Please check the preview https://honry.github.io/webnn-samples/image_classification/.

File: image_classification/resnet50v2_nchw.js [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    const conv3x3 = await this.buildConv_(
+        batchNorm1, '1', stageName, {padding: [1, 1, 1, 1], strides});
+    const batchNorm2 = await this.buildBatchNorm_(conv3x3, '2', stageName);
+    const conv1 = await this.buildConv_(batchNorm2, '2', stageName);
+    const conv2 = await this.buildConv_(batchNorm0, '3', stageName, {strides});
+
+    return this.builder_.add(conv1, conv2);
+  }
+
+  async buildLinearBottleneck2_(input, stageName, nameIndex) {
```
Comment (huningxin) [inline image_classification/resnet50v2_nchw.js:84]: According to [paper](https://arxiv.org/pdf/1512.03385.pdf) and [training script](https://github.com/onnx/models/blob/master/vision/classification/resnet/train_resnet.ipynb), I'd suggest to combine `buildLinearBottleneck1_` and `buildLinearBottleneck2_` into `buildBottleneckV2_` function such as
```js
async buildBottleneckV2_(input, stageName, nameIndex, downsample = false, stride = 1) {
    let residual = input;
    const strides = [stride, stride];
    const bn1 = await this.buildBatchNorm_(input, nameIndex, stageName);
    if (downsample) {
      residual = await this.buildConv_(bn1, nameIndex + 3, stageName, {strides});
    } else {
      nameIndex = nameIndex + 1;
    }
    const conv1 = await this.buildConv_(bn1, nameIndex, stageName);
    const bn2 = await this.buildBatchNorm_(conv1, nameIndex + 1, stageName);
    const conv2 = await this.buildConv_(bn2, nameIndex + 2, stageName, {padding: [1, 1, 1, 1], strides});
    const bn3 = await this.buildBatchNorm_(conv2, nameIndex + 2, stageName);
    const conv3 = await this.buildConv_(bn3, nameIndex + 3, stageName);

    return this.builder_.add(conv3, residual);
  }
```

There might be syntax errors in my code, but you got the idea. Thanks!

**Review by huningxin — COMMENTED** [review]
Thanks @Honry , I left some comments, please take a look.

File: image_classification/resnet50v2_nchw.js (hunk 1 above)
Comment (Honry) [reply image_classification/resnet50v2_nchw.js:84]:
```js
async buildBottlenectV2_(input, stageName, nameIndex, downsample = false, stride = 1) {
    let residual = input;
    const strides = [stride, stride];

    const batchNorm0 = await this.buildBatchNorm_(input, nameIndex, stageName);
    let conv0;
    if (downsample) {
      conv0 = await this.buildConv_(batchNorm0, nameIndex, stageName);
    } else {
      conv0 = await this.buildConv_(batchNorm0, nameIndex + 1, stageName);
    }
    const batchNorm1 = await this.buildBatchNorm_(conv0, nameIndex + 1, stageName);
    let conv3x3;
    if (downsample) {
      conv3x3 = await this.buildConv_(
        batchNorm1, nameIndex + 1, stageName, {padding: [1, 1, 1, 1], strides});
    } else {
      conv3x3 = await this.buildConv_(
        batchNorm1, nameIndex + 2, stageName, {padding: [1, 1, 1, 1]});
    }
    const batchNorm2 = await this.buildBatchNorm_(conv3x3, nameIndex + 2, stageName);
    if (downsample) {
      const conv1 = await this.buildConv_(batchNorm2, nameIndex + 2, stageName);
      const conv2 = await this.buildConv_(batchNorm0, nameIndex + 3, stageName, {strides});
      return this.builder_.add(conv1, conv2);
    } else {
      const conv1 = await this.buildConv_(batchNorm2, nameIndex + 3, stageName);
      return this.builder_.add(conv1, input);
    }
  }
```
@huningxin, I created a combined function to view how many different between these two `bottleneck`, it looks a bit complicated, especially for those un-aligned `nameIndex`.

Comment (huningxin) [reply image_classification/resnet50v2_nchw.js:84]: @Honry , I think my sample code is equivalent to yours for:



> ```js
>  if (downsample) {
>       conv0 = await this.buildConv_(batchNorm0, nameIndex, stageName);
>     } else {
>       conv0 = await this.buildConv_(batchNorm0, nameIndex + 1, stageName);
>     }
> ```

You can reset the base of `nameIndex` when `downsample` is false, e.g. `nameIndex = nameIndex + 1;`



> ```js
> if (downsample) {
>       conv3x3 = await this.buildConv_(
>         batchNorm1, nameIndex + 1, stageName, {padding: [1, 1, 1, 1], strides});
>     } else {
>       conv3x3 = await this.buildConv_(
>         batchNorm1, nameIndex + 2, stageName, {padding: [1, 1, 1, 1]});
>     }
> ```

You can make sure if `downsample` is false, the `stride` is always 1.



> ```js
> if (downsample) {
>       const conv1 = await this.buildConv_(batchNorm2, nameIndex + 2, stageName);
>       const conv2 = await this.buildConv_(batchNorm0, nameIndex + 3, stageName, {strides});
>       return this.builder_.add(conv1, conv2);
>     } else {
>       const conv1 = await this.buildConv_(batchNorm2, nameIndex + 3, stageName);
>       return this.builder_.add(conv1, input);
>     }
> ```

You can set the `residual = await this.buildConv_(bn1, nameIndex + 3, stageName, {strides});` when `downsample` is true, otherwise set the `residual = input`, then add it into the conv3 at the last as `return this.builder_.add(conv3, residual);`.

Comment (Honry) [reply image_classification/resnet50v2_nchw.js:84]: The problem is the `nameIndex` in common `const batchNorm1 = await this.buildBatchNorm_(conv0, nameIndex + 1, stageName);` and `const batchNorm2 = await this.buildBatchNorm_(conv3x3, nameIndex + 2, stageName);` declarations  could not be align when we set `nameIndex = nameIndex + 1;` in `downsample` is false.

Comment (Honry) [reply image_classification/resnet50v2_nchw.js:84]: Thus we need more `if...else...` for `batchNorm1` and `batchNorm2`.

Comment (huningxin) [reply image_classification/resnet50v2_nchw.js:84]: The architecture of residual unit is basically same with different shortcut connect (downsample or not). So it would make sense to use one code flow to build it. If the problem is only about the nameIndex, probably we can supply them, e.g.

```js
async buildBottleneckV2_(input, stageName, nameIndices, downsample = false, stride = 1) {
  let residual = input;
  const strides = [stride, stride];
  const bn1 = await this.buildBatchNorm_(input, nameIndices[0], stageName);
  const conv1 = await this.buildConv_(bn1, nameIndices[1], stageName);
  const bn2 = await this.buildBatchNorm_(conv1, nameIndices[2], stageName);
  const conv2 = await this.buildConv_(bn2, nameIndices[3], stageName, {padding: [1, 1, 1, 1], strides});
  const bn3 = await this.buildBatchNorm_(conv2, nameIndices[4], stageName);
  const conv3 = await this.buildConv_(bn3, nameIndices[5], stageName);
  if (downsample) {
    residual = await this.buildConv_(bn1, nameIndices[6], stageName, {strides});
  }
  return this.builder_.add(conv3, residual);
}

```

**Comment by Honry** [comment]
@huningxin, I've combined the `buildLinearBottleneck1_` and `buildLinearBottleneck2_` into `buildBottleneckV2_`, PTAL, thanks!

**Review by huningxin — APPROVED** [review]
LGTM. Thanks!
